python/python/bystro/supervised_ppca/gf_generative_pt.py

```diff
@@ -0,0 +1,715 @@
+            ]
+
+            sigma = torch.sum(list_covs,
+                dim=0,
+            )
+            WWT = torch.matmul(torch.transpose(W_, 0, 1), W_)
+            Sigma = WWT + torch.diag(sigma)
+
+            m = MultivariateNormal(torch.zeros(p), Sigma)
+
+            like_tot = torch.mean(m.log_prob(X_batch))
+            like_prior = _prior(trainable_variables)
+            posterior = like_tot + like_prior / N
+            loss = -1 * posterior
+
+            optimizer.zero_grad()
+            loss.backward()
+            optimizer.step()
+
+            self._save_losses(i, like_tot, like_prior, posterior)
+
+        self._store_instance_variables(trainable_variables)
+
+        return self
+
+    def get_covariance(self):
+        """
+        Gets the covariance matrix
+
+        Sigma = W^TW + sigma2*I
+
+        Parameters
+        ----------
+        None
+
+        Returns
+        -------
+        covariance : np.array-like(p,p)
+            The covariance matrix
+        """
+        covariance = np.dot(self.W_.T, self.W_) + np.diag(self.sigmas_)
+        return covariance
+
+    def _create_prior(self):
+        """
+        This creates the function representing prior on pararmeters
+
+        Parameters
+        ----------
+        log_prior : function
+            The function representing the negative log density of the prior
+        """
+        pd = self.prior_options
+
+        def log_prior(trainable_variables):
+            list_gamma_log_probs = []
+            for k in self.n_groups:
+                list_gamma_log_probs.append(
+                    Gamma(pd["alpha"], pd["beta"]).log_prob(
+                        trainable_variables[k + 1]
+                    )
+                )
+
+            gamma_log_probs = torch.stack(list_gamma_log_probs)
+            return torch.mean(gamma_log_probs)
+
+        return log_prior
+
+    def _fill_prior_options(self, prior_options):
+        """
+        Fills in options for prior parameters
+
+        Paramters
+        ---------
+        new_dict : dictionary
+            The prior parameters used to specify the prior
+        """
+        default_dict = {"alpha": 3.0, "beta": 3.0}
+        new_dict = {**default_dict, **prior_options}
+        return new_dict
+
+    def _initialize_variables(self, X):
+        """
+        Initializes the variables of the model
+
+        Parameters
+        ----------
+        X : np.array-like,(n_samples,p)
+            The data
+
+        Returns
+        -------
+        W_ : tf.Variable-like,(n_components,p)
+            The loadings of our latent factor model
+
+        sigmal_ : list
+            A list of the isotropic noises for each group
+        """
+        model = dp.PCA(self.n_components)
+        S_hat = model.fit_transform(X)
+        W_init = model.components_.astype(np.float32)
+        W_ = torch.tensor(W_init, requires_grad=True)
+        X_recon = np.dot(S_hat, W_init)
+        diff = np.mean((X - X_recon) ** 2)
```

Comment:
same comment as before about describing the initialization strategy

Reply:
This can't be placed in the init method unless we abandon the sklearn style of providing data in the fit method.